ring/replication_set_tracker.go

```diff
@@ -495,7 +495,14 @@ func (t *inflightInstanceTracker) addInstance(replicationSetIdx int, instance *I
 	t.mx.Lock()
 	defer t.mx.Unlock()
 
-	t.inflight[replicationSetIdx][instance] = struct{}{}
+	// Check if the instance has already been added.
+	for i, instances := 0, t.inflight[replicationSetIdx]; i < len(instances); i++ {
+		if instances[i] == instance {
```

Comment:
```suggestion
	for _, inflight := range t.inflight[replicationSetIdx] {
		if inflight == instance {
```

can we do this instead?

Reply:
Yes, we can. No impact on perf. Done in https://github.com/grafana/dskit/pull/498/commits/4cd5dbbe9ccebb0097799bb29d29ae929cc0de31